zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1221,21 +1238,85 @@ private void reconnect() {
     }
   }
 
+
+
+  private void doAsyncSync(final ZooKeeper zk, final String path, final long startT,
+      final ZkAsyncCallbacks.SyncCallbackHandler cb) {
+    zk.sync(path, cb,
+        new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, true) {
+          @Override
+          protected void doRetry() throws Exception {
+            doAsyncSync(zk, path, System.currentTimeMillis(), cb);
+          }
+        });
+  }
+
+  /*
+   *  Note, retrySync takes a ZooKeeper (client) object and pass it to doAsyncSync().
+   *  The reason we do this is that we want to ensure each new session event is preceded with exactly
```

Comment:
Exactly one zk.sync() for one new session if successful.  Otherwise, the retry would fail before the next "retrySync" is inserted to zkclient event queue.

Basically, for one retrySync(), it would only be invoked preceding the current new session. It would only retry connectionloss and would fail (thus not retry) if there is other exception. Thus, we don't need to worry this event clogging the event queue. Nor will any two of them run at the same time.